### PR TITLE
enable risk-affected-pipelines to register additional post-processors

### PR DESCRIPTION
**Pull request by rmudambi**
### Description

- *Category*: feature
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-7006

### Changes and notes
Enable risk-affected-pipelines to register additional post-processors
Remove duplicative risks/calibration_constant.py file that is here because of a failed merge conflict resolution

### Testing
Added new tests of feature
Ran make check
Confirmed that this worked in the MNCH NO model

### CODE REVIEW

**Comment by rmudambi** [comment]
PR Review: Enable risk-affected-pipelines to register additional post-processors
Summary
This commit adds an [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) parameter to [register_risk_affected_attribute_producer()](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [register_risk_affected_rate_producer()](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) in [calibration_constant.py](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). The new parameter accepts a single [AttributePostProcessor](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) or a sequence of them, which are applied to the target pipeline after the calibration constant adjustment. The parameter defaults to an empty tuple, maintaining backward compatibility. The test file was substantially rewritten to use a parameterized [TestProducer](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) class covering attribute/rate producers × no/single/multiple post-processors × calibration scenarios.

Design
Silent import shadowing creates a feature-gapped fork (High) — [transition.py:21-27](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [state.py:24-36](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [special_disease.py:23-31](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) all import register_risk_affected_*_producer from [causal_factor.calibration_constant](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and then immediately shadow it with an import from [risks.calibration_constant](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). The [risks](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) version lacks [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). Any future attempt to use the new parameter from these modules will raise TypeError at runtime. Additionally, [risks/__init__.py](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) re-exports the old version, so [from vivarium_public_health.risks import register_risk_affected_rate_producer](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) gives a stale API.

Suggestion: Delete risks/calibration_constant.py and replace [risks/__init__.py](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) exports with re-exports from [causal_factor.calibration_constant](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). Remove the shadowed imports from the three disease modules.

Normalization of [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) could be done earlier (Low) — [calibration_constant.py:161-164](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): The [isinstance(self._additional_post_processors, Sequence)](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) check is done in [setup()](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) rather than at the public function boundary. Normalizing to a list in the public functions and storing a [Sequence[AttributePostProcessor]](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) internally would simplify the class and eliminate the runtime branch.

[str](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) is a [Sequence](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) — theoretical hole (Info) — [calibration_constant.py:163](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): [isinstance("some_string", Sequence)](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) is True. If someone accidentally passes a string, it would be silently iterated character-by-character. The type annotation provides static protection, but runtime won't catch it. This matches the framework's convention, so it's low risk.

Maintainability
Docstring says "dynamic rate pipeline" in the attribute producer (Low) — [calibration_constant.py:54](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): The [source](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) parameter docstring for [register_risk_affected_attribute_producer](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) says "The source for the dynamic rate pipeline" — copy-pasted from the rate variant. Should say "attribute pipeline." Same issue in [calibration_constant.py:88](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) where the rate producer's docstring says "attribute producer" instead of "rate producer."

Tests import from [causal_factor](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) but live under [risks](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (Low) — [test_calibration_constant.py:25-28](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) imports from [vivarium_public_health.causal_factor.calibration_constant](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), but the test file's location suggests it tests the [risks](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) package. The public API surface via [risks.__init__](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (which re-exports from the stale copy) has zero test coverage. Consolidating the modules (per finding 1) would resolve this.

DRY
Entire module duplicated (~230 lines) (Critical) — risks/calibration_constant.py is a near-complete copy of causal_factor/calibration_constant.py, minus the new [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) feature. The two files are now functionally divergent. Any bug fix or feature must be applied to both.

Two public functions are near-identical (Medium) — [register_risk_affected_attribute_producer](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) ([calibration_constant.py:38-67](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) and [register_risk_affected_rate_producer](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) ([calibration_constant.py:70-100](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) share identical signatures, docstrings, and bodies, differing only by [is_rate=False](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) vs True. This is a pre-existing issue amplified by the commit (adding the parameter required changes in both). Consider extracting a shared private helper.

Tests
[test_calibration_constant_of_one_zeroes](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) doesn't test interaction with post-processors (High) — [test_calibration_constant.py:276](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): This test is not parameterized over [post_processor](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). When calibration=1 zeroes out the value and a post-processor like [_add_one_post_processor](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) is attached, the expected result is 1.0, not 0.0. This interaction is untested — if the calibration zero-out accidentally short-circuited subsequent post-processors, no test would catch it.

Suggestion: Parameterize over [post_processor](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and update expected values to [transform(0)](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html).

Only one post-processor ordering tested (Medium) — [test_calibration_constant.py:152-154](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): Only double_then_add_one ([v*2 + 1](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) is tested. Adding a reversed-order param (add_one_then_double → [(v+1)*2](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) would catch ordering regressions, since [v*2+1 ≠ (v+1)*2](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for non-trivial [v](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html).

Zero base value path untested (Medium) — [calibration_constant.py:233-237](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): [_apply_calibration_constant](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) has a [value[value != 0]](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) guard, but no test uses [base_value=0](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html). If the guard were removed or inverted, no test would fail.

Documentation
CHANGELOG typo (Low) — [CHANGELOG.rst:3](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): "Enable adding addition post-processors" should be "Enable adding additional post-processors."

Concept docs not updated (Low) — [calibration_constant.rst:27-40](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): The "How Calibration Works" section doesn't mention the new [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) capability. Consider adding a note that additional post-processors can be chained after the calibration step.

Functionality
Backward compatibility is preserved — The default [additional_post_processors=()](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) means all existing callers work unchanged. [isinstance((), Sequence)](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) → [list(())](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) → [] → final list [[self._apply_calibration_constant]](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), matching old behavior. No existing caller passes the new parameter.

Post-processor execution order is correct — [calibration_constant.py:172-175](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) builds [[self._apply_calibration_constant, *preferred_post_processor_list]](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), applying calibration first, then additional post-processors. This matches the documented contract.

No immediate runtime breakage from the dual-module issue — While finding 1/6 describes a serious architectural issue, no existing caller currently passes [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html), so the shadowed imports don't cause failures today. The issue is latent.

Minor Nits
CHANGELOG date mismatch — [CHANGELOG.rst:1](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html): Entry says 04/23/26 but the commit date is 2026-04-24.
Overall
The core feature — threading [additional_post_processors](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) through to the pipeline registration — is well-implemented and backward compatible. The test parameterization strategy is solid and provides good coverage of the new parameter's type variations (empty tuple, single callable, list of callables).

The critical concern is the duplicated [risks/calibration_constant.py](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) module (findings 1, 6) and the shadowed imports in three disease modules. While this is a pre-existing architectural issue not introduced by this commit, the commit widens the divergence between the two copies. This should be addressed either in this PR or as an immediate follow-up to prevent the [risks](vscode-file://vscode-app/c:/Users/rmudambi/AppData/Local/Programs/Microsoft%20VS%20Code/10c8e557c8/resources/app/out/vs/code/electron-browser/workbench/workbench.html) version from becoming a confusing trap for future developers.

Secondary priorities are strengthening test coverage around post-processor ordering (9), the zero-out + post-processor interaction (8), and the zero base value edge case (10).

**Review by albrja — APPROVED** [review]

File: tests/causal_factor/test_calibration_constant.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+from vivarium.framework.values import AttributePostProcessor, ValuesManager
+
+from tests.test_utilities import build_table_with_age
+from vivarium_public_health.causal_factor.calibration_constant import (
+    get_calibration_constant_pipeline_name,
+    register_risk_affected_attribute_producer,
+    register_risk_affected_rate_producer,
+)
+from vivarium_public_health.population import BasePopulation
+
```
Comment (stevebachmeier) [inline tests/causal_factor/test_calibration_constant.py:30]: I'm assuming this is a direct copy/paste from the `risks/test_calibration_constant.py`?

**Review by stevebachmeier — APPROVED** [review]

File: tests/causal_factor/test_calibration_constant.py (hunk 1 above)
Comment (rmudambi) [reply tests/causal_factor/test_calibration_constant.py:30]: Yes
